x-pack/plugins/uptime/public/components/monitor/synthetics/waterfall/components/waterfall_chart.tsx

```diff
@@ -0,0 +1,221 @@
+import { EuiThemeProvider } from '../../../../../../../observability/public';
+import { useWaterfallContext } from '../context/waterfall_chart';
+import {
+  WaterfallChartOuterContainer,
+  WaterfallChartFixedTopContainer,
+  WaterfallChartFixedTopContainerSidebarCover,
+  WaterfallChartFixedAxisContainer,
+  WaterfallChartSidebarContainer,
+  WaterfallChartSidebarContainerInnerPanel,
+  WaterfallChartSidebarContainerFlexGroup,
+  WaterfallChartSidebarFlexItem,
+  WaterfallChartChartContainer,
+  WaterfallChartLegendContainer,
+  WaterfallChartTooltip,
+} from './styles';
+import { WaterfallData } from '../types';
+
+const Tooltip = ({ header }: TooltipInfo) => {
+  const { data, renderTooltipItem } = useWaterfallContext();
+  const relevantItems = data.filter((item) => {
+    return item.x === header?.value;
+  });
+  return (
+    <WaterfallChartTooltip>
+      <EuiFlexGroup direction="column" gutterSize="none">
+        {relevantItems.map((item, index) => {
+          return (
+            <EuiFlexItem key={index}>{renderTooltipItem(item.config.tooltipProps)}</EuiFlexItem>
+          );
+        })}
+      </EuiFlexGroup>
+    </WaterfallChartTooltip>
+  );
+};
+
+export type RenderItem<I = any> = (item: I, index: number) => JSX.Element;
+
+interface WaterfallChartProps {
+  tickFormat: TickFormatter;
+  domain: DomainRange;
+  barStyleAccessor: BarStyleAccessor;
+  renderSidebarItem?: RenderItem;
+  renderLegendItem?: RenderItem;
+  maxHeight?: number;
+}
+
+const getUniqueBars = (data: WaterfallData) => {
+  return data.reduce<number[]>((acc, item) => {
+    if (acc.indexOf(item.x) === -1) {
+      acc.push(item.x);
+      return acc;
+    } else {
+      return acc;
+    }
+  }, []);
+};
+
+const BAR_HEIGHT = 32;
+const getChartHeight = (data: WaterfallData): number => getUniqueBars(data).length * BAR_HEIGHT;
+
+const MAIN_GROW_SIZE = 8;
+const SIDEBAR_GROW_SIZE = 2;
+
+export const WaterfallChart = ({
+  tickFormat,
+  domain,
+  barStyleAccessor,
+  renderSidebarItem,
+  renderLegendItem,
+  maxHeight = 600,
+}: WaterfallChartProps) => {
+  const { data, sidebarItems, legendItems } = useWaterfallContext();
+  const [darkMode] = useUiSetting$<boolean>('theme:darkMode');
+
+  const theme = useMemo(() => {
+    return darkMode ? EUI_CHARTS_THEME_DARK.theme : EUI_CHARTS_THEME_LIGHT.theme;
+  }, [darkMode]);
+
+  return (
+    <EuiThemeProvider darkMode={darkMode}>
+      <WaterfallChartOuterContainer height={maxHeight}>
+        <>
+          <WaterfallChartFixedTopContainer>
+            <EuiFlexGroup gutterSize="none">
+              {sidebarItems && sidebarItems.length > 0 && (
+                <EuiFlexItem grow={SIDEBAR_GROW_SIZE}>
+                  <WaterfallChartFixedTopContainerSidebarCover paddingSize="none" />
+                </EuiFlexItem>
+              )}
+              <EuiFlexItem grow={sidebarItems && sidebarItems.length > 0 ? MAIN_GROW_SIZE : true}>
```

Comment:
i think we can abstract this component into two pieces, side bar and main chart component, WDYT?

Totally optional thing.